File: ZWSCountdown/Classes/ZWSCountdownUtils.m
//
//  ZWSCountdownUtils.m
//  ZWSCountdown
//
//  Created by zhaowensky on 2016/12/28.
//  Copyright © 2016年 ZWS. All rights reserved.
//

#import "ZWSCountdownUtils.h"
#import <UIKit/UIKit.h>

@interface ZWSCountdownUtils ()
@property (nonatomic,strong) NSTimer                 *timer;
@property (nonatomic,copy  ) ZWSCountdownUtilsHandle handle;
@property (nonatomic,copy  ) NSString                *phoneNumber;
@property (nonatomic,copy  ) NSString                *business;

@property (nonatomic,assign) int                     currentSecond;
@property (nonatomic,assign) int                     totalSecond;

@end

@implementation ZWSCountdownUtils

#pragma makr - init
+ (instancetype)getCountdownUtils
{
    return [[ZWSCountdownUtils alloc]init];
}

-(id)init
{
    self = [super init];
    if(self){
        [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(didEnterBackground) name:UIApplicationDidEnterBackgroundNotification object:nil];
        [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(didBecomeActive) name:UIApplicationDidBecomeActiveNotification object:nil];
    }
    return self;
}

-(void)dealloc
{
    [[NSNotificationCenter defaultCenter]removeObserver:self];
}

#pragma mark - application
-(void)didEnterBackground
{
    [self saveCountdown:_phoneNumber business:_business second:_currentSecond];
}

-(void)didBecomeActive
{
    [self getCountDown];
}

#pragma mark - control
-(void)startCountdown:(NSString*)phoneNumber
             business:(NSString*)business
               second:(int)second
             callback:(ZWSCountdownUtilsHandle)handle
{
    self.phoneNumber = phoneNumber;
    self.business = business;
    self.totalSecond = second;
    
    NSAssert(phoneNumber, @"phoneNumber is nil.");
    NSAssert(business, @"business is nil.");
    NSAssert(!(second == 0), @"second > 0");
    
    [self getCountDown];
    self.handle = handle;
    self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(updateTimer) userInfo:nil repeats:YES];
}

-(void)stopCountdown
{
    if(self.timer){
        [self.timer invalidate];
        self.timer = nil;
    }
    if(self.currentSecond == 0){
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        NSString *key = [NSString stringWithFormat:@"%@_%@",_phoneNumber,_business];
        [defaults removeObjectForKey:key];
        [defaults synchronize];
    }else{
        [self saveCountdown:_phoneNumber business:_business second:_currentSecond];
    }
}

-(void)updateTimer
{
    if(self.handle){
        self.handle(self.currentSecond);
        if(self.currentSecond == 0){
            [self.timer invalidate];
            self.timer = nil;
        }else{
            self.currentSecond --;
        }
    }
}

-(BOOL)checkCountdown:(NSString *)phoneNumber
             business:(NSString *)business
{
    return [self getCountDownTime:phoneNumber business:business] > 0;
}

#pragma mark - set/get time
-(void)saveCountdown:(NSString*)phoneNumber
            business:(NSString*)business
              second:(int)second
{
    NSDictionary *countValue = @{@"second":[NSNumber numberWithInt:second],
                                 @"saveDate":[NSDate date]};
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *key = [NSString stringWithFormat:@"%@_%@",_phoneNumber,_business];
    [defaults setObject:countValue forKey:key];
    [defaults synchronize];
}

-(void)getCountDown
{
    int result = [self getCountDownTime:_phoneNumber business:_business];
    if(result == 0){
        self.currentSecond = self.totalSecond;
    }else{
        self.currentSecond = result;
    }
}

-(int)getCountDownTime:(NSString*)phone business:(NSString*)business
{
    int result = 0;
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *key = [NSString stringWithFormat:@"%@_%@",phone,business];
    NSDictionary *countValue = [defaults objectForKey:key];
    NSDate *saveDate = [countValue objectForKey:@"saveDate"];
    if(saveDate){
        int second = [[countValue objectForKey:@"second"] intValue];
        NSTimeInterval timeInterval = [[NSDate date] timeIntervalSinceDate:saveDate];
        if(timeInterval >= second){
            result = 0;
        }else{
            result = second - timeInterval;
        }
    }
    return result;
}


@end

File: ZWSCountdown/ZWSCountdown.Swift/Classes/ZWSCountdownUtils.swift
//
//  ZWSCountdownUtils.swift
//  ZWSCountdown.Swift
//
//  Created by zhaowensky on 2017/1/5.
//  Copyright © 2017年 ZWS. All rights reserved.
//

import Foundation

typealias ZWSCountdownUtilsHandle = (Int32) -> ()

public final class ZWSCountdownUtils: NSObject{
    //MARK: - prams
    private var timer:Timer!
    private var phoneNumber:String!
    private var business:String!
    private var currentSecond:Int32 = 0
    private var totalSecond:Int32 = 0
    private var handle:ZWSCountdownUtilsHandle?
    
    private let nameEnterBackground = Notification.Name.UIApplicationDidEnterBackground;
    private let nameBecomeActive = Notification.Name.UIApplicationDidBecomeActive
    
    //MARK: - init
    public override init() {
        super.init()
        
        NotificationCenter.default.addObserver(self, selector:#selector(didEnterBackground(notification:)), name: nameEnterBackground, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(didBecomeActive(notification:)), name: nameBecomeActive, object: nil)
    }
    
    deinit{
        NotificationCenter.default.removeObserver(self);
    }
    
    //MARK: - private
    @objc fileprivate func didEnterBackground(notification:Notification) -> Void {
        saveCountdown(phoneNumber: phoneNumber, business: business, second: currentSecond);
    }

    @objc fileprivate func didBecomeActive(notification:Notification) -> Void {
        getCountDown()
    }
    
    @objc fileprivate func updateTimer() -> Void {
        if (handle != nil){
            handle!(currentSecond)
            if (currentSecond == 0){
                timer.invalidate()
            }else{
                currentSecond -= 1
            }
        }
    }
    
    //TODO: - open
    
    /// 开启倒计时 start
    ///
    /// - Parameters:
    ///   - phoneNumber: 电话号码 phoneNumber
    ///   - business: 业务代号（eg:com.xxx.xxx）
    ///   - second: 倒计时长 time
    ///   - callback: block countdown time
    func startCountdown(startPhone:String?,startBusiness:String?,second:Int32,
                        callback:ZWSCountdownUtilsHandle?) -> Void {
        phoneNumber = startPhone
        business = startBusiness
        totalSecond = second
        
//        assert(startPhone == nil, "startPhone is nil")
//        assert(startBusiness == nil, "business is nil")
//        assert(second == 0, "totalSecond > 0")
        
        getCountDown()
        handle = callback
        timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(updateTimer), userInfo: nil, repeats: true)
    }
    
    func stopCountdown() -> Void {
        if (timer != nil){
            timer.invalidate()
        }
        if(currentSecond == 0){
            let defaults = UserDefaults.standard
            let key = phoneNumber + "_" + business
            defaults.removeObject(forKey: key)
            defaults.synchronize()
        }else{
            saveCountdown(phoneNumber: phoneNumber, business: business, second: currentSecond)
        }
    }
    
    func checkCountdown(phoneNumber:String,business:String) -> Bool {
        return getCountDownTime(phoneNumber: phoneNumber, business: business) > 0
    }
    
    //MARK: - setter / getter time
    private func saveCountdown(phoneNumber:String,business:String,second:Int32) -> Void {
        let key = phoneNumber + "_" + business
        let countValue:[String:Any] = ["second":NSNumber.init(value: second),"saveDate":Date.init()]
        let defaults = UserDefaults.standard
        defaults.set(countValue, forKey: key)
        defaults.synchronize()
    }
    
    private func getCountDown() -> Void {
        let result = getCountDownTime(phoneNumber:phoneNumber as String, business:business as String)
        if result == 0 {
            currentSecond = totalSecond
        }else{
            currentSecond = Int32(result)
        }
    }
    
    private func getCountDownTime(phoneNumber:String,business:String) -> Int32 {
        var result:Double = 0
        let key = phoneNumber + "_" + business
        
        let defaults = UserDefaults.standard
        var countValue = defaults.dictionary(forKey:key)
        let saveValue = countValue?["saveDate"] as? Date
        if (saveValue) != nil {
            let second = (countValue?["second"] as! NSNumber).doubleValue
            let timeInterval = Date().timeIntervalSince(saveValue!)
            if timeInterval >= second {
                result = 0
            }else{
                result = second - timeInterval
            }
        }
        return Int32(result)
    }
}






